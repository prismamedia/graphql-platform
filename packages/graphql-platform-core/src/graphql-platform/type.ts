import { GraphQLInputType, GraphQLOutputType } from 'graphql';
import { Resource } from './resource';

export * from './type/input';
export * from './type/output';

export enum TypeKind {
  Input,
  Output,
}

interface Type<
  TTypeKind extends TypeKind,
  TType extends GraphQLInputType | GraphQLOutputType,
> {
  kind: TTypeKind;
  id: string;
  resource: Resource;
  isSupported(): boolean;
  getGraphQLType(): TType;
}

export interface InputType extends Type<TypeKind.Input, GraphQLInputType> {}

export interface OutputType extends Type<TypeKind.Output, GraphQLOutputType> {}
