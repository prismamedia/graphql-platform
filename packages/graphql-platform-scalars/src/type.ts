import * as utils from '@prismamedia/graphql-platform-utils';
import * as graphql from 'graphql';
import type { ValueOf } from 'type-fest';
import { GraphQLBigInt } from './type/big-int.js';
import { GraphQLDateTime } from './type/date-time.js';
import { GraphQLDate } from './type/date.js';
import { GraphQLDraftJS } from './type/draft-js.js';
import { GraphQLEmailAddress } from './type/email-address.js';
import { jsonTypesByName } from './type/json.js';
import { GraphQLNonEmptyString } from './type/non-empty-string.js';
import { GraphQLNonEmptyTrimmedString } from './type/non-empty-trimmed-string.js';
import { GraphQLUnsignedBigInt } from './type/unsigned-bigint.js';
import { GraphQLUnsignedFloat } from './type/unsigned-float.js';
import { GraphQLUnsignedInt } from './type/unsigned-int.js';
import { GraphQLURL } from './type/url.js';
import { uuidTypesByName } from './type/uuid.js';

export * from './type/big-int.js';
export * from './type/date-time.js';
export * from './type/date.js';
export * from './type/draft-js.js';
export * from './type/email-address.js';
export * from './type/json.js';
export * from './type/non-empty-string.js';
export * from './type/non-empty-trimmed-string.js';
export * from './type/unsigned-bigint.js';
export * from './type/unsigned-float.js';
export * from './type/unsigned-int.js';
export * from './type/url.js';
export * from './type/uuid.js';

/**
 * @see https://graphql.org/graphql-js/type/#scalars
 */
export const basicTypesByName = {
  Boolean: graphql.GraphQLBoolean,
  Float: graphql.GraphQLFloat,
  ID: graphql.GraphQLID,
  Int: graphql.GraphQLInt,
  String: graphql.GraphQLString,
} satisfies Record<string, graphql.GraphQLScalarType>;

export const typesByName = Object.freeze({
  // Basic
  ...basicTypesByName,

  // GraphQL Platform
  BigInt: GraphQLBigInt,
  Date: GraphQLDate,
  DateTime: GraphQLDateTime,
  DraftJS: GraphQLDraftJS,
  EmailAddress: GraphQLEmailAddress,
  NonEmptyString: GraphQLNonEmptyString,
  NonEmptyTrimmedString: GraphQLNonEmptyTrimmedString,
  UnsignedBigInt: GraphQLUnsignedBigInt,
  UnsignedFloat: GraphQLUnsignedFloat,
  UnsignedInt: GraphQLUnsignedInt,
  URL: GraphQLURL,
  ...jsonTypesByName,
  ...uuidTypesByName,
} satisfies Record<string, graphql.GraphQLScalarType>);

type TypesByName = typeof typesByName;

export type TypeName = keyof TypesByName;
export const typeNames = Object.keys(typesByName);

export type Type = ValueOf<TypesByName>;
export const types = Object.values(typesByName);

export type GetInternalValueByType<TType extends Type> =
  TType extends graphql.GraphQLScalarType<infer TInternalValue, any>
    ? TInternalValue
    : never;

export type GetExternalValueByType<TType extends Type> =
  TType extends graphql.GraphQLScalarType<any, infer TExternalValue>
    ? TExternalValue
    : never;

export function getTypeByName(
  maybeTypeName: TypeName,
  path?: utils.Path,
): Type {
  const scalarType = typesByName[maybeTypeName];
  if (!scalarType) {
    throw new utils.UnexpectedValueError(
      `a scalar type name among "${typeNames.join(', ')}"`,
      maybeTypeName,
      { path },
    );
  }

  return scalarType;
}

export function assertType(
  maybeType: unknown,
  path?: utils.Path,
): asserts maybeType is Type {
  if (!types.includes(maybeType as any)) {
    throw new utils.UnexpectedValueError(
      `a scalar type among "${typeNames.join(', ')}"`,
      maybeType,
      { path },
    );
  }
}

export function ensureType(maybeType: unknown, path: utils.Path): Type {
  assertType(maybeType, path);

  return maybeType;
}
