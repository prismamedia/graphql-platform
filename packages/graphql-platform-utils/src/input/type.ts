import * as graphql from 'graphql';
import { UnexpectedValueError } from '../error.js';
import type { Path } from '../path.js';
import {
  getGraphQLNamedInputType,
  isNamedInputType,
  isNamedInputTypePublic,
  NamedInputType,
  parseNamedInputLiteral,
  parseNamedInputValue,
  validateNamedInputType,
} from './type/named.js';
import { isWrappingInputType, WrappingInputType } from './type/wrapping.js';

export * from './type/named.js';
export * from './type/wrapping.js';

export type InputType = NamedInputType | WrappingInputType;

export function isInputType(
  maybeInputType: unknown,
): maybeInputType is InputType {
  return (
    isNamedInputType(maybeInputType) || isWrappingInputType(maybeInputType)
  );
}

export function assertInputType(
  maybeInputType: unknown,
  path?: Path,
): asserts maybeInputType is InputType {
  if (!isInputType(maybeInputType)) {
    throw new UnexpectedValueError(`an input type`, maybeInputType, { path });
  }
}

export function ensureInputType(
  maybeInputType: unknown,
  path?: Path,
): InputType {
  assertInputType(maybeInputType, path);

  return maybeInputType;
}

export function isInputTypePublic(type: InputType): boolean {
  return isNamedInputType(type)
    ? isNamedInputTypePublic(type)
    : isInputTypePublic(type.ofType);
}

export function validateInputType(type: InputType): void {
  isNamedInputType(type)
    ? validateNamedInputType(type)
    : validateInputType(type.ofType);
}

export function getGraphQLInputType(type: InputType): graphql.GraphQLInputType {
  return isNamedInputType(type)
    ? getGraphQLNamedInputType(type)
    : type.getGraphQLInputType();
}

export function parseInputValue(
  type: InputType,
  value: unknown,
  path?: Path,
): any {
  return isNamedInputType(type)
    ? parseNamedInputValue(type, value, path)
    : type.parseValue(value, path);
}

export type NonNullNonVariableGraphQLValueNode = Exclude<
  graphql.ValueNode,
  graphql.NullValueNode | graphql.VariableNode
>;

export function parseInputLiteral(
  type: InputType,
  value: graphql.ValueNode,
  variableValues?: graphql.GraphQLResolveInfo['variableValues'],
  path?: Path,
): any {
  if (value.kind === graphql.Kind.NULL) {
    return parseInputValue(type, null, path);
  } else if (value.kind === graphql.Kind.VARIABLE) {
    return parseInputValue(type, variableValues?.[value.name.value], path);
  }

  return isNamedInputType(type)
    ? parseNamedInputLiteral(type, value, variableValues, path)
    : type.parseLiteral(value, variableValues, path);
}

export function getNamedInputType(inputType: InputType): NamedInputType {
  return isWrappingInputType(inputType)
    ? getNamedInputType(inputType.ofType)
    : inputType;
}
