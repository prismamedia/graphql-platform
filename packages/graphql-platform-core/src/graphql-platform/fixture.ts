import { Entries, fromEntries } from '@prismamedia/graphql-platform-utils';
import { DepGraph } from 'dependency-graph';
import { OperationDefinitionNode, print } from 'graphql';
import { Logger } from 'winston';
import { GraphQLPlatform, GraphQLRequest } from '../graphql-platform';
import { CreateOneDataInputValue } from './operation';
import { AnyResource, Relation, SerializedFieldValue } from './resource';
import { TypeKind, WhereUniqueInputValue } from './type';

export class FixtureGraph extends DepGraph<Fixture> {}

export type FixtureFieldData = SerializedFieldValue;

export type FixtureRelationData = null | Fixture['name'];

export type FixtureComponentData = FixtureFieldData | FixtureRelationData;

export type FixtureData = {
  [componentName: string]: FixtureComponentData;
};

export type FixtureDataMap = {
  [fixtureName: string]: FixtureData;
};

export class Fixture {
  readonly logger?: Logger;
  protected id?: WhereUniqueInputValue;

  public constructor(
    readonly name: string,
    readonly resource: AnyResource,
    readonly data: FixtureData,
    protected graph: FixtureGraph,
    protected gp: GraphQLPlatform,
  ) {
    this.logger = gp.getLogger();
  }

  public toString(): string {
    return `${this.resource}.${this.name}`;
  }

  public getRelatedFixture(relation: Relation): Fixture | null {
    this.resource.getRelationMap().assert(relation);

    const relationValue = this.data[relation.name];

    if (relationValue == null) {
      if (relation.isRequired()) {
        throw new Error(`In the fixture "${this.name}", the relation "${relation}" can't be null.`);
      }
    } else if (!(typeof relationValue === 'string' && this.graph.hasNode(relationValue))) {
      throw new Error(
        `In the fixture "${this.name}", the relation "${relation}" has to reference an "${relation.getTo()}" fixture.`,
      );
    }

    return relationValue ? this.graph.getNodeData(relationValue) : null;
  }

  public getCreateMutationAST(): OperationDefinitionNode {
    const operationDefinitionNode: OperationDefinitionNode = {
      kind: 'OperationDefinition',
      operation: 'mutation',
      variableDefinitions: [
        {
          kind: 'VariableDefinition',
          variable: {
            kind: 'Variable',
            name: {
              kind: 'Name',
              value: 'data',
            },
          },
          type: {
            kind: 'NonNullType',
            type: {
              kind: 'NamedType',
              name: {
                kind: 'Name',
                value: this.resource.getMutation('CreateOne').getDataType().name,
              },
            },
          },
        },
      ],
      selectionSet: {
        kind: 'SelectionSet',
        selections: [
          {
            kind: 'Field',
            name: {
              kind: 'Name',
              value: this.resource.getMutation('CreateOne').name,
            },
            alias: {
              kind: 'Name',
              value: 'id',
            },
            arguments: [
              {
                kind: 'Argument',
                name: {
                  kind: 'Name',
                  value: 'data',
                },
                value: {
                  kind: 'Variable',
                  name: {
                    kind: 'Name',
                    value: 'data',
                  },
                },
              },
            ],
            selectionSet: this.resource
              .getFirstPublicUnique()
              .getSelectionNode(TypeKind.Output)
              .toAST(),
          },
        ],
      },
    };

    return operationDefinitionNode;
  }

  public getCreateMutationSource(): string {
    return print(this.getCreateMutationAST());
  }

  public getCreateMutationVariables(): GraphQLRequest['variableValues'] {
    return {
      data: fromEntries([
        // Fields
        ...[...this.resource.getFieldSet()].map(field =>
          !field.isFullyManaged() ? [field.name, this.data[field.name] as FixtureFieldData] : undefined,
        ),
        // Relations
        ...[...this.resource.getRelationSet()].map(relation => {
          if (!relation.isFullyManaged()) {
            const relatedFixture = this.getRelatedFixture(relation);
            if (relatedFixture) {
              return [relation.name, { connect: relatedFixture.assertId() }];
            }
          }
        }),
      ] as Entries<CreateOneDataInputValue>),
    };
  }

  public getCreateMutation(): GraphQLRequest {
    return {
      source: this.getCreateMutationSource(),
      variableValues: this.getCreateMutationVariables(),
    };
  }

  public async load(contextValue?: GraphQLRequest['contextValue']): Promise<void> {
    try {
      const { id } = await this.gp.execute<{ id: WhereUniqueInputValue }>({
        ...this.getCreateMutation(),
        contextValue,
      });

      this.id = this.resource.getInputType('WhereUnique').assert(id);
    } catch (error) {
      this.logger && this.logger.error(error);

      throw error;
    }
  }

  public assertId(): WhereUniqueInputValue {
    if (!this.id) {
      throw new Error(`The fixture "${this.name}" is not loaded yet.`);
    }

    return this.id;
  }
}
