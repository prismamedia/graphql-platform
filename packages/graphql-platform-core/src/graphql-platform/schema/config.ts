import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
