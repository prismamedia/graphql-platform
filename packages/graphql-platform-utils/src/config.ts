export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
