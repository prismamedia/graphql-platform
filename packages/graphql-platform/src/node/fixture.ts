import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import type { ConnectorInterface } from '../connector-interface.js';
import type { Node, NodeValue } from '../node.js';
import type { Seeding } from '../seeding.js';
import type { UniqueConstraintValue } from './definition.js';
import { EdgeCreationInputAction } from './type/input/creation/field.js';

export type NodeFixtureReference = string;

export type NodeFixtureData = utils.PlainObject;

export class NodeFixture<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
> {
  public readonly dependencies: ReadonlySet<NodeFixtureReference>;

  public constructor(
    public readonly seeding: Seeding<TRequestContext, TConnector>,
    public readonly node: Node<TRequestContext, TConnector>,
    public readonly reference: NodeFixtureReference,
    public readonly data: NodeFixtureData,
  ) {
    if (typeof reference !== 'string') {
      throw new utils.UnexpectedValueError('a string', reference);
    }

    if (!utils.isPlainObject(data)) {
      throw new utils.UnexpectedValueError('a plain-object', data);
    }

    this.dependencies = new Set(
      Array.from(this.node.edgesByName.values()).reduce<NodeFixtureReference[]>(
        (dependencies, edge) => {
          const maybeEdgeReference = data[edge.name];
          if (maybeEdgeReference != null) {
            dependencies.push(maybeEdgeReference);
          }

          return dependencies;
        },
        [],
      ),
    );
  }

  @Memoize()
  public get dependents(): ReadonlySet<NodeFixtureReference> {
    return new Set(
      Array.from(this.seeding.fixturesByReference.values()).reduce<
        NodeFixtureReference[]
      >((dependents, fixture) => {
        if (fixture.dependencies.has(this.reference)) {
          dependents.push(fixture.reference);
        }

        return dependents;
      }, []),
    );
  }

  @Memoize((context: TRequestContext) => context)
  public async load(context: TRequestContext): Promise<NodeValue> {
    const data = this.node.creationInputType.parseValue(
      Object.fromEntries(
        await Promise.all(
          Object.entries(this.data).map(
            async ([componentName, componentValue]) => [
              componentName,
              this.node.edgesByName.has(componentName)
                ? {
                    [EdgeCreationInputAction.CONNECT]:
                      await this.seeding.fixturesByReference
                        .get(componentValue)!
                        .getIdentifier(context),
                  }
                : componentValue,
            ],
          ),
        ),
      ),
      utils.addPath(undefined, this.reference),
    ) as any;

    return this.node
      .getMutationByKey('create-one')
      .execute({ data, selection: this.node.selection }, context) as any;
  }

  @Memoize((context: TRequestContext) => context)
  public async getIdentifier(
    context: TRequestContext,
  ): Promise<UniqueConstraintValue> {
    const nodeValue = await this.load(context);

    return this.node.identifier.parseValue(nodeValue);
  }
}
