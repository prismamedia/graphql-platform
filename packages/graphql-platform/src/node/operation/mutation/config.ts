import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
