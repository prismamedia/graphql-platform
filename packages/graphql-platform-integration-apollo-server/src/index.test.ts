import { ApolloServer } from '@apollo/server';
import { afterAll, beforeAll, describe, expect, it } from '@jest/globals';
import type { InMemoryBroker } from '@prismamedia/graphql-platform';
import {
  createMyGP,
  type MyGP,
} from '@prismamedia/graphql-platform-connector-mariadb/__tests__/config.js';
import { myAdminContext } from '@prismamedia/graphql-platform/__tests__/config.js';
import * as fixtures from '@prismamedia/graphql-platform/__tests__/fixture.js';
import { ApolloServerIntegration } from './index.js';

describe('GraphQL-Platform Integration Apollo Server', () => {
  let gp: MyGP<InMemoryBroker>;

  beforeAll(async () => {
    gp = createMyGP('integration_apollo_server', {
      subscription: { public: true },
    });

    await gp.connector.setup();
    await gp.seed(myAdminContext, fixtures.constant);
  });

  afterAll(() => gp.connector.teardown());

  it('creates an ApolloServer instance', async () => {
    const server = new ApolloServerIntegration(gp);
    expect(server).toBeInstanceOf(ApolloServer);

    await expect(
      server.executeOperation(
        {
          query: `query GetArticles($first: UnsignedInt!) {
            articles(orderBy: [createdAt_ASC], first: $first) {
              status
              title
              body
              views
              score
              machineTags
              metas
              tagCount
              tags (orderBy: [order_ASC], first: 10) {
                order
                tag {
                  title
                }
              }
              createdAt
            }
          }`,
          variables: {
            first: 10,
          },
        },
        { contextValue: myAdminContext },
      ),
    ).resolves.toMatchSnapshot();

    await expect(
      server.executeOperation(
        {
          query: `subscription SubscribeToArticleChanges {
            articleChanges(where: { status: PUBLISHED }) {
              ... on ArticleDeletion {
                id
              }
              ... on Article {
                id
                status
                title
              }
            }
          }`,
        },
        { contextValue: myAdminContext },
      ),
    ).resolves.toMatchSnapshot();
  });
});
