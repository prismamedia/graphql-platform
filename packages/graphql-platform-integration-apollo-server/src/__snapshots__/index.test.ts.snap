// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Platform Integration Apollo Server creates an ApolloServer instance 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "articles": [
          {
            "body": {
              "blocks": [],
              "entityMap": {},
            },
            "createdAt": "2022-01-01T00:00:00.000Z",
            "machineTags": null,
            "metas": {
              "myArbitraryKey": "with my arbitrary value",
            },
            "score": 0.5,
            "status": "DRAFT",
            "tagCount": 0,
            "tags": [],
            "title": "My first draft article",
            "views": "0",
          },
          {
            "body": null,
            "createdAt": "2022-02-01T00:00:00.000Z",
            "machineTags": null,
            "metas": null,
            "score": 0.5,
            "status": "DRAFT",
            "tagCount": 0,
            "tags": [],
            "title": "My second draft article",
            "views": "0",
          },
          {
            "body": null,
            "createdAt": "2022-03-01T00:00:00.000Z",
            "machineTags": null,
            "metas": null,
            "score": 0.5,
            "status": "PUBLISHED",
            "tagCount": 2,
            "tags": [
              {
                "order": 0,
                "tag": {
                  "title": "TV",
                },
              },
              {
                "order": 1,
                "tag": {
                  "title": "High-tech",
                },
              },
            ],
            "title": "My first published article",
            "views": "0",
          },
          {
            "body": {
              "blocks": [],
              "entityMap": {},
            },
            "createdAt": "2022-04-01T00:00:00.000Z",
            "machineTags": [
              "namespace:key=a_value",
              "namespace:key=other_value",
            ],
            "metas": {
              "aKey": "withAnyValue",
            },
            "score": 0.12,
            "status": "PUBLISHED",
            "tagCount": 1,
            "tags": [
              {
                "order": 0,
                "tag": {
                  "title": "Fashion",
                },
              },
            ],
            "title": "My second published article",
            "views": "1234567890",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`GraphQL-Platform Integration Apollo Server creates an ApolloServer instance 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "articles": [
          {
            "body": {
              "blocks": [],
              "entityMap": {},
            },
            "createdAt": "2022-01-01T00:00:00.000Z",
            "machineTags": null,
            "metas": {
              "myArbitraryKey": "with my arbitrary value",
            },
            "score": 0.5,
            "status": "DRAFT",
            "tagCount": 0,
            "tags": [],
            "title": "My first draft article",
            "views": "0",
          },
          {
            "body": null,
            "createdAt": "2022-02-01T00:00:00.000Z",
            "machineTags": null,
            "metas": null,
            "score": 0.5,
            "status": "DRAFT",
            "tagCount": 0,
            "tags": [],
            "title": "My second draft article",
            "views": "0",
          },
          {
            "body": null,
            "createdAt": "2022-03-01T00:00:00.000Z",
            "machineTags": null,
            "metas": null,
            "score": 0.5,
            "status": "PUBLISHED",
            "tagCount": 2,
            "tags": [
              {
                "order": 0,
                "tag": {
                  "title": "TV",
                },
              },
              {
                "order": 1,
                "tag": {
                  "title": "High-tech",
                },
              },
            ],
            "title": "My first published article",
            "views": "0",
          },
          {
            "body": {
              "blocks": [],
              "entityMap": {},
            },
            "createdAt": "2022-04-01T00:00:00.000Z",
            "machineTags": [
              "namespace:key=a_value",
              "namespace:key=other_value",
            ],
            "metas": {
              "aKey": "withAnyValue",
            },
            "score": 0.12,
            "status": "PUBLISHED",
            "tagCount": 1,
            "tags": [
              {
                "order": 0,
                "tag": {
                  "title": "Fashion",
                },
              },
            ],
            "title": "My second published article",
            "views": "1234567890",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
